File: hs/cgen_ex/c/Animal.cpp
#define CGEN_OUTPUT_INTERN
#include "Animal.h"
void Animals_Animal_delete(Animal* this_ptr)
{
    delete this_ptr;
}

void Animals_Animal_make_sound(Animal* this_ptr)
{
    this_ptr->make_sound();
}

int Animals_Animal_get_age(const Animal* this_ptr)
{
    return this_ptr->get_age();
}

void Animals_Animal_increment_age(Animal* this_ptr)
{
    this_ptr->increment_age();
}

File: hs/cgen_ex/hs/Animal.hs
{-# LANGUAGE ForeignFunctionInterface #-}
module Animal(
animal_delete, 
animal_make_sound, 
animal_get_age, 
animal_increment_age
)

where

import Types
import Control.Monad

import Foreign
import Foreign.C.String
import Foreign.C.Types

foreign import ccall "Animal.h Animals_Animal_delete" c_animal_delete :: Animal -> IO ()
animal_delete :: Animal -> IO ()
animal_delete p1 =   c_animal_delete p1

foreign import ccall "Animal.h Animals_Animal_make_sound" c_animal_make_sound :: Animal -> IO ()
animal_make_sound :: Animal -> IO ()
animal_make_sound p1 =   c_animal_make_sound p1

foreign import ccall "Animal.h Animals_Animal_get_age" c_animal_get_age :: Animal -> IO CInt
animal_get_age :: Animal -> IO Int
animal_get_age p1 =  liftM fromIntegral $  c_animal_get_age p1

foreign import ccall "Animal.h Animals_Animal_increment_age" c_animal_increment_age :: Animal -> IO ()
animal_increment_age :: Animal -> IO ()
animal_increment_age p1 =   c_animal_increment_age p1

